File: day2/task1.py
def is_safe(r: list[int]) -> bool:
    if len(r) <= 1:
        return True
    if len(r) == 2:
        return 1 <= abs(r[0] - r[1]) <= 3

    for x1, x2, x3 in zip(r, r[1:], r[2:]):
        if not (
            (x1 < x2 < x3 or x1 > x2 > x3)
            and 1 <= abs(x1 - x2) <= 3
            and 1 <= abs(x2 - x3) <= 3
        ):
            return False
    return True


with open("input.txt") as f:
    reports = [list(map(int, l.split(" "))) for l in f.readlines()]

res = sum(is_safe(r) for r in reports)
print(res)

File: day2/task1.hs
readline :: String -> [Int]
readline "" = []
readline s =
  let (n, s') = head (reads s :: [(Int, String)])
   in (n : readline s')

readlines :: [String] -> [[Int]]
readlines = foldr ((:) . readline) []

cond :: Int -> Int -> Bool
cond a b = (\x -> x >= 1 && x <= 3) (abs (a - b))

monot :: [Int] -> Bool
monot [] = True
monot [_] = True
monot [x1, x2] = cond x1 x2
monot (x1 : x2 : x3 : xs) =
  (x1 < x2 && x2 < x3 || x1 > x2 && x2 > x3)
    && cond x1 x2
    && cond x2 x3
    && monot (x2 : x3 : xs)

main = do
  contents <- readFile "input.txt"
  let l = readlines $ lines contents
  print $ foldl (\acc xs -> if monot xs then acc + 1 else acc) 0 l
